src/components/TodoItem.tsx

```diff
@@ -0,0 +1,192 @@
+/* eslint-disable @typescript-eslint/no-unused-expressions */
+import React, {
+  useContext, useEffect, useState,
+  FormEvent, KeyboardEvent, useRef,
+} from 'react';
+import classNames from 'classnames';
+import { Todo } from '../types/Todo';
+import { TodoContext } from '../contexts/TodoContext';
+import { deleteTodo, updateTodo } from '../api/todos';
+import { Error } from '../types/Error';
+
+interface Props {
+  todo: Todo;
+}
+
+type EditEvent = FormEvent<HTMLFormElement>
+| KeyboardEvent<HTMLInputElement> | null;
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
+  const {
+    setTodos,
+    setErrorMessage,
+    loadingTodo,
+    setLoadingTodo,
+  } = useContext(TodoContext);
+
+  const { id, title, completed } = todo;
+  const [isEditing, setIsEditing] = useState(false);
+  const [newTitle, setNewTitle] = useState(title);
+
+  const myInput = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    myInput.current && myInput.current.focus();
+  }, [isEditing, myInput]);
+
+  const handleEditStart = () => {
+    setIsEditing(true);
+  };
+
+  const handleEditCancel = () => {
+    setIsEditing(false);
+    setNewTitle(title);
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setLoadingTodo(todo);
+
+    deleteTodo(todoId)
+      .then(() => setTodos((prev) => prev.filter((t) => t.id !== todoId)))
+      .catch(() => setErrorMessage(Error.Delete))
+      .finally(() => setLoadingTodo(null));
+  };
+
+  const handleChangeCheckbox = (todoId: number) => {
+    setLoadingTodo(todo);
+
+    updateTodo(todoId, { completed: !completed })
+      .then(() => setTodos(prev => prev
+        .map(newTodo => (newTodo.id === todoId
+          ? ({ ...newTodo, completed: !completed })
+          : newTodo))))
+      .catch(() => setErrorMessage(Error.Update))
+      .finally(() => {
+        setLoadingTodo(null);
+      });
+  };
```

Comment:
the same with update
![image](https://github.com/mate-academy/react_todo-app-with-api/assets/91946036/ee3ca2a1-f063-4096-858f-b8c7ec92d7d6)



check other api too

Reply:
I've managed the Delete function, but I don't know how to optimize the function for updates.  I would be very grateful if you could help me figure out how to rewrite it. I've wasted a lot of time trying to figure it out.  🥲